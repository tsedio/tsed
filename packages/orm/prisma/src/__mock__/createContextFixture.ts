import {TransformContext} from "../generator/domain/TransformContext.js";

export function createContextFixture(): TransformContext {
  return {
    modelsMap: new Map()
      .set("User", {
        name: "User",
        fields: [{type: "Role"}]
      })
      .set("Role", {
        name: "Role",
        fields: [{type: "User"}, {type: "Transitive"}]
      })
      .set("Transitive", {
        name: "Transitive",
        fields: [{type: "String"}, {type: "User"}]
      })
      .set("Hello", {
        name: "Hello",
        fields: [{type: "String"}, {type: "User"}]
      })
  } as any;
}
