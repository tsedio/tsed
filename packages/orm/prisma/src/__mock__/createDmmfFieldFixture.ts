import {DmmfField} from "../generator/domain/DmmfField";

export interface FieldFixtureOptions {
  kind: string;
  name: string;
  isRequired: boolean;
  isNullable: boolean;
  type: string;
  isList: boolean;
  documentation: string;
  isInputType: boolean;
}

export function createDmmfFieldFixture(options: Partial<FieldFixtureOptions> = {}) {
  return new DmmfField({
    field: {
      name: "user",
      isRequired: false,
      type: "User",
      isList: false,
      ...options
    },
    schemaArg: {
      ...options
    },
    model: {
      isInputType: options.isInputType,
      addImportDeclaration: jest.fn()
    }
  });
}
