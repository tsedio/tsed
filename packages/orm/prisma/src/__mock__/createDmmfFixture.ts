export function createDmmfFixture() {
  return require("./dmmf.json");
}

export function createDmmfWithTypesFixture() {
  return require("./dmmf.types.json");
}
