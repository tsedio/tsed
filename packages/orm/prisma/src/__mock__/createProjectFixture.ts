import filedirname from "filedirname";
import {join} from "path";
import {ModuleKind, Project, ScriptTarget} from "ts-morph";

// FIXME remove when esm is ready
const [, rootDir] = filedirname();

const SNAPSHOT_DIR = `${rootDir}/../../test/snapshots`;

export function createProjectFixture(dir = "/") {
  const baseDir = join(SNAPSHOT_DIR, dir);
  const project = new Project({
    compilerOptions: {
      target: ScriptTarget.ES2019,
      module: ModuleKind.CommonJS,
      emitDecoratorMetadata: true,
      experimentalDecorators: true,
      esModuleInterop: true
    }
  });

  return {
    project,
    baseDir,
    render(file: string) {
      const absolutePath = join(baseDir, file);
      const sourceFile = project.getSourceFile(absolutePath);

      if (!sourceFile) {
        throw `File not found: ${file}`;
      }

      sourceFile.formatText({indentSize: 2});
      const actualValue = sourceFile.getFullText();

      return {
        ...expect(actualValue),
        toMatchSnapshot() {
          expect(actualValue).toMatchSnapshot();
        }
      };
    }
  };
}
