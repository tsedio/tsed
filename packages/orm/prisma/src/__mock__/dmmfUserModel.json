{
  "model": {
    "name": "User",
    "isEmbedded": false,
    "dbName": null,
    "fields": [
      {
        "name": "id",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": true,
        "isReadOnly": false,
        "type": "Int",
        "hasDefaultValue": true,
        "default": {
          "name": "autoincrement",
          "args": []
        },
        "isGenerated": false,
        "isUpdatedAt": false,
        "documentation": "@TsED.Groups(\"!creation\")\nComment"
      },
      {
        "name": "createdAt",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "DateTime",
        "hasDefaultValue": true,
        "default": {
          "name": "now",
          "args": []
        },
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "email",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": true,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false,
        "documentation": "@TsED.Email()\n@TsED.Description(\"User email. This email must be unique!\")"
      },
      {
        "name": "weight",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Float",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "is18",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Boolean",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "name",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "successorId",
        "kind": "scalar",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": true,
        "type": "Int",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "successor",
        "kind": "object",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "User",
        "hasDefaultValue": false,
        "relationName": "BlogOwnerHistory",
        "relationFromFields": ["successorId"],
        "relationToFields": ["id"],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "predecessor",
        "kind": "object",
        "isList": false,
        "isRequired": false,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "User",
        "hasDefaultValue": false,
        "relationName": "BlogOwnerHistory",
        "relationFromFields": [],
        "relationToFields": [],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "role",
        "kind": "enum",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Role",
        "hasDefaultValue": true,
        "default": "USER",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "posts",
        "kind": "object",
        "isList": true,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Post",
        "hasDefaultValue": false,
        "relationName": "PostToUser",
        "relationFromFields": [],
        "relationToFields": [],
        "relationOnDelete": "NONE",
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "keywords",
        "kind": "scalar",
        "isList": true,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "String",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      },
      {
        "name": "biography",
        "kind": "scalar",
        "isList": false,
        "isRequired": true,
        "isUnique": false,
        "isId": false,
        "isReadOnly": false,
        "type": "Json",
        "hasDefaultValue": false,
        "isGenerated": false,
        "isUpdatedAt": false
      }
    ],
    "isGenerated": false,
    "idFields": [],
    "uniqueFields": [],
    "uniqueIndexes": []
  },
  "modelType": {
    "name": "User",
    "fields": [
      {
        "name": "id",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Int",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "createdAt",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "DateTime",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "email",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "weight",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Float",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "is18",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Boolean",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "name",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "successorId",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "Int",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "successor",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "User",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": false
        }
      },
      {
        "name": "predecessor",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "User",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": false
        }
      },
      {
        "name": "role",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Role",
          "namespace": "model",
          "location": "enumTypes",
          "isList": false
        }
      },
      {
        "name": "posts",
        "args": [
          {
            "name": "where",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostWhereInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "orderBy",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostOrderByWithRelationInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": true
              },
              {
                "type": "PostOrderByWithRelationInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "cursor",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostWhereUniqueInput",
                "namespace": "prisma",
                "location": "inputObjectTypes",
                "isList": false
              }
            ]
          },
          {
            "name": "take",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "Int",
                "location": "scalar",
                "isList": false
              }
            ]
          },
          {
            "name": "skip",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "Int",
                "location": "scalar",
                "isList": false
              }
            ]
          },
          {
            "name": "distinct",
            "isRequired": false,
            "isNullable": false,
            "inputTypes": [
              {
                "type": "PostScalarFieldEnum",
                "namespace": "prisma",
                "location": "enumTypes",
                "isList": true
              }
            ]
          }
        ],
        "isNullable": true,
        "outputType": {
          "type": "Post",
          "namespace": "model",
          "location": "outputObjectTypes",
          "isList": true
        }
      },
      {
        "name": "keywords",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "String",
          "location": "scalar",
          "isList": true
        }
      },
      {
        "name": "biography",
        "args": [],
        "isNullable": false,
        "outputType": {
          "type": "Json",
          "location": "scalar",
          "isList": false
        }
      },
      {
        "name": "_count",
        "args": [],
        "isNullable": true,
        "outputType": {
          "type": "UserCountOutputType",
          "namespace": "prisma",
          "location": "outputObjectTypes",
          "isList": false
        }
      }
    ]
  },
  "isInputType": false
}
