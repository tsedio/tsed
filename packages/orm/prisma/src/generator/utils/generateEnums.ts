import {DMMF} from "@prisma/generator-helper";
import {Project} from "ts-morph";
import {toMap} from "@tsed/core";
import {DmmfEnum} from "../domain/DmmfEnum";
import path from "path";
import {generateDocuments} from "./generateDocuments";
import {transformEnumsToEnums} from "../transform/transformEnumsToEnums";
import {generateOutputsBarrelFile} from "./generateOutputsBarrelFile";

export function generateEnums(dmmf: DMMF.Document, project: Project, baseDirPath: string): boolean {
  const enumsMap = toMap<string, DMMF.DatamodelEnum>(dmmf.datamodel.enums, "name");
  const enums = DmmfEnum.getEnums(dmmf, enumsMap);
  const enumsDirPath = path.resolve(baseDirPath, "enums");
  const enumsDirectory = project.createDirectory(enumsDirPath);
  const enumsIndex = enumsDirectory.createSourceFile(`index.ts`, undefined, {overwrite: true});

  const exportedEnums = generateDocuments<DmmfEnum>(enums, enumsDirectory, (document) => transformEnumsToEnums(document));

  if (enums.length) {
    return !!generateOutputsBarrelFile(enumsIndex, exportedEnums);
  }

  return false;
}
