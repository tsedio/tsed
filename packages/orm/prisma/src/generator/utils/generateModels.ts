import {DMMF} from "@prisma/generator-helper";
import {getValue, toMap} from "@tsed/core";
import path from "path";
import {Project} from "ts-morph";

import {DmmfModel} from "../domain/DmmfModel.js";
import {transformModelToClass} from "../transform/transformModelToClass.js";
import {generateDocuments} from "./generateDocuments.js";
import {generateOutputsBarrelFile} from "./generateOutputsBarrelFile.js";

export function generateModels(dmmf: DMMF.Document, project: Project, baseDirPath: string) {
  const modelsMap = toMap<string, DMMF.Model>(getValue(dmmf, "datamodel.models", []), "name");
  const typesMap = toMap<string, DMMF.Model>(getValue(dmmf, "datamodel.types", []), "name");

  const models = DmmfModel.getModels(dmmf, modelsMap, typesMap);
  const modelsDirPath = path.resolve(baseDirPath, "models");
  const modelsDirectory = project.createDirectory(modelsDirPath);
  const modelsIndex = modelsDirectory.createSourceFile(`index.ts`, undefined, {overwrite: true});

  const exportedModels = generateDocuments<DmmfModel>(models, modelsDirectory, (model, sourceFile) => {
    return transformModelToClass(model, {
      dmmf,
      modelsMap
    });
  });

  generateOutputsBarrelFile(modelsIndex, exportedModels);
}
