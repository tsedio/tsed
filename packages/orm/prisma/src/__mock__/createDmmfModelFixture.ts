import {DmmfModel} from "../generator/domain/DmmfModel.js";
import dmmfUserModel from "./dmmfUserModel.json";

export function createDmmfModelFixture() {
  return new DmmfModel(dmmfUserModel as any);
}
