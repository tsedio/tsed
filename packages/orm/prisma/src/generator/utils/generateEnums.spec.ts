import {createProjectFixture} from "../../__mock__/createProjectFixture";
import {generateModels} from "./generateModels";
import {createDmmfFixture} from "../../__mock__/createDmmfFixture";
import {generateEnums} from "./generateEnums";

describe("generateEnums", () => {
  it("should generate filese", () => {
    const {project, render, baseDir} = createProjectFixture("generate_enums");
    const dmmf = createDmmfFixture();

    generateEnums(dmmf, project, baseDir);

    render("/enums/Role.ts").toMatchSnapshot();
  });
});
