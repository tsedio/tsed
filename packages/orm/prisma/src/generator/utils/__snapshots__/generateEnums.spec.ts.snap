// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generateEnums > should generate filese 1`] = `
"
export enum Role {
  USER = "USER",
  ADMIN = "ADMIN"
}
"
`;

exports[`generateEnums should generate filese 1`] = `
"
export enum Role {
  USER = \\"USER\\",
  ADMIN = \\"ADMIN\\"
}
"
`;
